File: src/main/java/de/hinundhergestellt/jhuh/service/ready2order/ArtooDataStore.java
package de.hinundhergestellt.jhuh.service.ready2order;

import de.hinundhergestellt.jhuh.core.task.Futures;
import de.hinundhergestellt.jhuh.vendors.ready2order.ArtooProduct;
import de.hinundhergestellt.jhuh.vendors.ready2order.ArtooProductClient;
import de.hinundhergestellt.jhuh.vendors.ready2order.ArtooProductGroup;
import de.hinundhergestellt.jhuh.vendors.ready2order.ArtooProductGroupClient;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.beans.factory.config.BeanDefinition;
import org.springframework.context.annotation.Scope;
import org.springframework.core.task.AsyncTaskExecutor;
import org.springframework.data.util.Lazy;
import org.springframework.stereotype.Service;

import java.util.List;

import static java.util.stream.Stream.concat;

@Service
@Scope(BeanDefinition.SCOPE_PROTOTYPE)
public class ArtooDataStore {

    private final Lazy<List<ArtooMappedCategory>> rootCategories;

    public ArtooDataStore(ArtooProductGroupClient productGroupClient,
                          ArtooProductClient productClient,
                          @Qualifier("applicationTaskExecutor") AsyncTaskExecutor taskExecutor) {
        var groups = taskExecutor.submit(() -> productGroupClient.findAll().toList());
        var products = taskExecutor.submit(() -> productClient.findAll().toList());
        rootCategories = Lazy.of(() -> new DataStoreBuilder(Futures.get(groups), Futures.get(products)).rootCategories());
    }

    public List<ArtooMappedCategory> getRootCategories() {
        return rootCategories.get();
    }

    @SuppressWarnings("ClassCanBeRecord")
    private static class DataStoreBuilder {

        private final List<ArtooProductGroup> groups;
        private final List<ArtooProduct> products;

        DataStoreBuilder(List<ArtooProductGroup> groups, List<ArtooProduct> products) {
            this.groups = groups;
            this.products = products;
        }

        List<ArtooMappedCategory> rootCategories() {
            return groups.stream()
                    .filter(it -> it.getParent() == 0)
                    .map(this::category)
                    .toList();
        }

        private ArtooMappedCategory category(ArtooProductGroup group) {
            var children = groups.stream()
                    .filter(it -> it.getTypeId() == 7)
                    .filter(it -> it.getParent() == group.getId())
                    .map(this::category)
                    .toList();

            var productsWithVariations = groups.stream()
                    .filter(it -> it.getTypeId() == 3)
                    .filter(it -> it.getParent() == group.getId())
                    .map(this::product);
            var singleProducts = products.stream()
                    .filter(it -> it.getProductGroupId() == group.getId())
                    .map(this::product);
            var products = concat(productsWithVariations, singleProducts).toList();

            return new ArtooMappedCategory(group, children, products);
        }

        private ArtooMappedProduct product(ArtooProductGroup group) {
            var variations = products.stream()
                    .filter(it -> it.getProductGroupId() == group.getId())
                    .map(ArtooMappedVariation::new)
                    .toList();

            return new GroupArtooMappedProduct(group, variations);
        }

        private ArtooMappedProduct product(ArtooProduct product) {
            return new SingleArtooMappedProduct(product);
        }
    }
}

File: src/main/java/de/hinundhergestellt/jhuh/service/ready2order/ArtooDataStore.kt
package de.hinundhergestellt.jhuh.service.ready2order

import de.hinundhergestellt.jhuh.vendors.ready2order.ArtooProduct
import de.hinundhergestellt.jhuh.vendors.ready2order.ArtooProductClient
import de.hinundhergestellt.jhuh.vendors.ready2order.ArtooProductGroup
import de.hinundhergestellt.jhuh.vendors.ready2order.ArtooProductGroupClient
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.core.task.AsyncTaskExecutor
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Service
import java.util.concurrent.Callable
import java.util.concurrent.Future
import java.util.concurrent.TimeUnit

@Service
class ArtooDataStore(
    private val productGroupClient: ArtooProductGroupClient,
    private val productClient: ArtooProductClient,
    @Qualifier("applicationTaskExecutor") private val taskExecutor: AsyncTaskExecutor
) {
    private lateinit var rootCategoriesFuture: Future<List<ArtooMappedCategory>>
    val rootCategories: List<ArtooMappedCategory>
        get() = rootCategoriesFuture.get()

    val stateChangeListeners = mutableListOf<() -> Unit>()

    init {
        refresh()
    }

    fun findAllProducts() =
        rootCategories.asSequence().flatMap { it.findAllProducts() }

    fun findProductById(id: Int) =
        rootCategories.firstNotNullOfOrNull { it.findProductById(id) }

    fun findCategoriesByProduct(product: ArtooMappedProduct) =
        rootCategories.asSequence().flatMap { it.findCategoriesByProduct(product) }

    @Scheduled(initialDelay = 15, fixedDelay = 15, timeUnit = TimeUnit.MINUTES)
    fun refresh() {
        rootCategoriesFuture = fetchRootCategories()
    }

    private fun fetchRootCategories(): Future<List<ArtooMappedCategory>> {
        val groupsFuture = taskExecutor.submit(Callable { productGroupClient.findAll().toList() })
        return taskExecutor.submit(Callable {
            val products = productClient.findAll().toList()
            val groups = groupsFuture.get()
            DataStoreBuilder(groups, products).rootCategories
                .also { taskExecutor.submit { stateChangeListeners.forEach { it() } } }
        })
    }
}

private class DataStoreBuilder(
    private val groups: List<ArtooProductGroup>,
    private val products: List<ArtooProduct>
) {
    val rootCategories =
        groups.asSequence()
            .filter { it.parent == null && it.typeId == 7 }
            .map { it.toCategory() }
            .toList()

    private fun ArtooProductGroup.toCategory(): ArtooMappedCategory {
        val children = groups.asSequence()
            .filter { it.parent == id && it.typeId == 7 }
            .map { it.toCategory() }
            .toList()
        val products = products.asSequence()
            .filter { it.productGroupId == id && it.baseId == null }
            .map { it.toProduct() }
            .toList()
        return ArtooMappedCategory(this, children, products)
    }

    private fun ArtooProduct.toProduct(): ArtooMappedProduct {
        val variations = products.asSequence()
            .filter { it.baseId == id }
            .map { ArtooMappedVariation(this, it) }
            .toList()
        return ArtooMappedProduct(this, variations)
    }
}
